compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/GeneratedFileCache.kt

```diff
@@ -0,0 +1,238 @@
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.io.Serializable
+import java.security.MessageDigest
+
+internal typealias MD5String = String
+
+internal class GeneratedFileCache private constructor(
+  private val binaryFile: File,
+  private val projectDir: ProjectDir,
+) : Serializable, Closeable {
+
+  private val generatedToContent = mutableMapOf<RelativeFile, String>()
+  private val generatedToSources = SimpleMultimap<RelativeFile, RelativeFile>()
+  private val sourcesToGenerated = SimpleMultimap<RelativeFile, RelativeFile>()
+
+  private val filesToMd5 = mutableMapOf<RelativeFile, MD5String>()
+
+  val sourceFiles: Set<RelativeFile> get() = sourcesToGenerated.keys
+
+  fun isGenerated(file: FileType): Boolean {
+    return when (file) {
+      is AbsoluteFile -> isGenerated(file.relativeTo(projectDir))
+      is RelativeFile -> generatedToContent.containsKey(file)
+    }
+  }
+
+  fun getSourceFiles(generatedFile: FileType): Set<RelativeFile> {
+    return when (generatedFile) {
+      is AbsoluteFile -> getSourceFiles(generatedFile.relativeTo(projectDir))
+      is RelativeFile -> generatedToSources[generatedFile]
+    }
+  }
+
+  fun getContent(generatedFile: FileType): String {
+    return when (generatedFile) {
+      is AbsoluteFile -> getContent(generatedFile.relativeTo(projectDir))
+      is RelativeFile -> generatedToContent.getValue(generatedFile)
+    }
+  }
+
+  fun getGeneratedFiles(sourceFile: FileType): Set<RelativeFile> {
+    return when (sourceFile) {
+      is AbsoluteFile -> getGeneratedFiles(sourceFile.relativeTo(projectDir))
+      is RelativeFile -> sourcesToGenerated[sourceFile]
+    }
+  }
+
+  fun getGeneratedFilesRecursive(sourceFile: FileType): Set<RelativeFile> {
+    return when (sourceFile) {
+      is AbsoluteFile -> getGeneratedFilesRecursive(sourceFile.relativeTo(projectDir))
+      is RelativeFile -> {
+        val visited = mutableSetOf<RelativeFile>()
+        generateSequence(sourcesToGenerated[sourceFile]) { generated ->
+          generated.filter { visited.add(it) }
+            .flatMapTo(mutableSetOf()) { sourcesToGenerated[it] }
+            .takeIf { it.iterator().hasNext() }
+        }
+          .flatten()
+          .toSet()
+      }
+    }
+  }
+
+  fun addSourceFile(sourceFile: RelativeFile) {
+    addMd5(sourceFile, overwrite = true)
+  }
+
+  private fun addMd5(sourceFile: RelativeFile, overwrite: Boolean) {
+    if (overwrite) {
+      filesToMd5[sourceFile] = sourceFile.md5()
+    } else {
+      filesToMd5.computeIfAbsent(sourceFile) { sourceFile.md5() }
+    }
+  }
+
+  fun addGeneratedFile(generated: GeneratedFileWithSources) {
+
+    val generatedAbsolute = AbsoluteFile(generated.file)
+    val generatedRelative = generatedAbsolute.relativeTo(projectDir)
+
+    generatedToContent[generatedRelative] = generated.content
+
+    addMd5(generatedRelative, overwrite = true)
+
+    filesToMd5.computeIfAbsent(generatedRelative) { generatedRelative.md5() }
```

Comment:
It looks like the second md5 line is a duplicate/no-op now

```suggestion
    addMd5(generatedRelative, overwrite = true)
```

Reply:
I blame CoPilot?

Nice catch.